### DISCUSSION TITLE
Drawing canvas on bitmap loses all color transform

### DISCUSSION THREAD

**Discussion by NB-NutBoi** [HaxeFlixel/flixel#2488] [Q&A]
- Haxe version: 4.1.5
- Flixel version:  4.9.0
- OpenFL version: 9.1.0
- Lime version: 7.9.0
- Affected targets: Windows, Mac

Tried to search everywhere for a solution but closest i got was a transparency issue, when i draw the camera canvas onto a new bitmap, all color transform info gets lost, i've tried every solution i could come up with and i only really found to turn blit rendering on, which is really slow, and to modify the FlxDrawQuadsItem class to directly apply the color transform to the bitmap rather than storing it in a shader, but that leads to ping-pong rendering if improperly done and is really slow, bringing performance down significantly, is there a better way to do this or is it an issue with openfl/flixel itself?
________________________________

sample state to reproduce the issue

```haxe
package;

import flixel.FlxG;
import flixel.FlxSprite;
import flixel.FlxState;
import flixel.util.FlxColor;
import openfl.display.BitmapData;
import openfl.display.PNGEncoderOptions;
import openfl.utils.ByteArray;
import sys.io.File;

class BrokenDraw extends FlxState
{
	override public function create()
	{
		super.create();

		var brokenBox = new FlxSprite(100, 100).makeGraphic(100, 100, FlxColor.WHITE);
		brokenBox.color = FlxColor.RED;
		add(brokenBox);
	}

	override public function update(elapsed:Float)
	{
		if (FlxG.keys.justPressed.F12)
		{
			var bitmap:BitmapData = new BitmapData(640, 480);
			bitmap.draw(camera.canvas);
			var b:ByteArray = new ByteArray();
			b = bitmap.encode(bitmap.rect, new PNGEncoderOptions(true), b);
			File.saveBytes("testImage.png", b);
		}
	}
}
```

________________________________

Observed behavior: The square in the game window is red, and the square in the generated .png is white

Expected behavior: Both squares should be red.

**Comment by Geokureli** [comment]
@NB-NutBoi I converted this to a discussion

The fact that camera.canvas is exposed seems unintended. this is not meant to be a "flixel feature" and, to my knowledge we have no intentions on adding support for this. Also i doubt this would work on all targets anyway

However `bitmapData.draw` has a colorTransform arg. so you may just need to do do bitmap.draw( camera.canvas, null, brokenBox.colorTransform);

**Reply by Geokureli** [reply]
I do think, if anything, this may be an OpenFL issue rather than a flixel issue. but I'm not too familiar with flixel rendering

**Reply by NB-NutBoi** [reply]
it does have a colorTransform arg, yes, but that just transforms the colors of the graphic that's being drawn, rather than transforming parts of the graphic to match the ingame transforms. 
Trust me when i say i've tried absolutely everything to find a way around this or fix the issue, but if you're using the color properties of the FlxSprite class, i found no reliable way to get an accurate bitmap of the screen in tile rendering, this was a last resort thing to see if someone had a fix or workaround here that does not involve modifying the renderer.

i should mention that this is also an issue with FlxScreenGrab from the flixel addons plugin, so you can also replicate the issue that way.
